runtime/strategies/create-handle-group.js

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2017 Google Inc. All rights reserved.
+// This code may only be used under the BSD style license found at
+// http://polymer.github.io/LICENSE.txt
+// Code distributed by Google as part of this project is also
+// subject to an additional IP rights grant found at
+// http://polymer.github.io/PATENTS.txt
+
+import {Strategy} from '../../strategizer/strategizer.js';
+import Recipe from '../recipe/recipe.js';
+import RecipeWalker from '../recipe/walker.js';
+import Handle from '../recipe/handle.js';
+
+export default class CreateHandleGroup extends Strategy {
+
+  async generate(inputParams) {
+    return Recipe.over(this.getResults(inputParams), new class extends RecipeWalker {
+      onRecipe(recipe) {
+        // Resolve constraints before assuming connections are free.
+        if (recipe.connectionConstraints.length > 0) return;
+
+        const freeConnections = recipe.handleConnections.filter(hc => !hc.handle && !hc.isOptional);
+        let maximalGroup = null;
+        for (let writer of freeConnections.filter(hc => hc.isOutput)) {
+          let compatibleConnections = [writer];
+          let involvedParticles = new Set([writer.particle]);
+
+          let foundSomeReader = false;
+          for (let reader of freeConnections.filter(hc => hc.isInput && hc !== writer)) {
+            if (Handle.effectiveType(null, [reader, ...compatibleConnections])
+                && !involvedParticles.has(reader.particle)) {
+              compatibleConnections.push(reader);
+              involvedParticles.add(reader.particle);
+              foundSomeReader = true;
+            }
+          }
+
+          // Only make a 'create' group for a writer->reader case.
+          if (!foundSomeReader) continue;
+
+          for (let otherWriter of freeConnections.filter(hc => hc.isOutput && !compatibleConnections.includes(hc))) {
```

Comment:
can't this be part of the above loop? set foundSomeReader depending on whether it is an input or not.

Reply:
It would have different semantics - right now it prioritizes the broadest set of readers before adding a new writer. The latter would narrow down what can be read, possibly to no fields at all.

If you have a writer of * {Text a} and a writer of * {Text b}, they can write to the same handle, but no one can read anything from it except for (maybe) presence of an entity.

The 'connects the biggest groups available' test would fail if you do that.